HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+        self.u = lambda c: CRRAutility(c, self.CRRA)
+        self.uP = lambda c: CRRAutilityP(c, self.CRRA)
+        self.u_inv = lambda u: CRRAutility_inv(u, self.CRRA)
+        self.uP_inv = lambda uP: CRRAutilityP_inv(uP, self.CRRA)
+
+    def create_params_namespace(self):
+
+        self.params = SimpleNamespace(
+            Discount=self.DiscFac * self.LivPrb,
+            CRRA=self.CRRA,
+            Rfree=self.Rfree,
+            PermGroFac=self.PermGroFac,
+        )
+
+    def define_boundary_constraint(self):
+
+        # Calculate the minimum allowable value of money resources in this period
+        self.BoroCnstNat = (
+            self.solution_next.attrs["m_nrm_min"] - 1
+        ) / self.params.Rfree
+
+        # if natural borrowing constraint is binding constraint, then we can not
+        # evaluate the value function at that point, so we must fill out the data
+        if self.BoroCnstArt is None or self.BoroCnstArt <= self.BoroCnstNat:
+            self.m_nrm_min = self.BoroCnstNat
+            self.nat_boro_cnst = True
+
+            self.borocnst = xr.Dataset(
+                coords={"mNrm": self.m_nrm_min, "aNrm": self.m_nrm_min},
+                data_vars={
+                    "cNrm": 0.0,
+                    "v": -np.inf,
+                    "v_inv": 0.0,
+                    "reward": -np.inf,
+                    "marginal_reward": np.inf,
+                    "v_der": np.inf,
+                    "v_der_inv": 0.0,
+                },
+            )
+
+        elif self.BoroCnstArt > self.BoroCnstNat:
+            self.m_nrm_min = self.BoroCnstArt
+            self.nat_boro_cnst = False
+
+            self.borocnst = xr.Dataset(
+                coords={"mNrm": self.m_nrm_min, "aNrm": self.m_nrm_min},
+                data_vars={"cNrm": 0.0},
+            )
+
+    def create_post_state(self):
+
+        if self.nat_boro_cnst:
+            # don't include natural borrowing constraint
+            a_grid = self.aXtraGrid + self.m_nrm_min
+        else:
+            # include artificial borrowing constraint
+            a_grid = np.append(0.0, self.aXtraGrid) + self.m_nrm_min
+
+        aVec = xr.DataArray(
+            a_grid,
+            name="aNrm",
+            dims=("aNrm"),
+            attrs={"long_name": "savings", "state": True},
+        )
+        post_state = xr.Dataset({"aNrm": aVec})
+
+        self.post_state = post_state
+
+    def state_transition(self, s=None, a=None, params=None):
+        """
+        state to post_state transition
+        """
+        ps = {}  # pytree
+        ps["aNrm"] = s["mNrm"] - a["cNrm"]
+        return ps
+
+    def post_state_transition(self, ps=None, params=None):
+        """
+        post_state to next_state transition
+        """
+        ns = {}  # pytree
+        ns["mNrm"] = ps["aNrm"] * params.Rfree / params.PermGroFac + 1
+        return ns
+
+    def reverse_transition(self, ps=None, a=None, params=None):
+        """state from post_state and actions"""
```

Comment:
'reverse' or 'inverse' ?

Reply:
I think reverse is more appropriate. State transition is forward (state to post-state, given action), reverse transition is backward (post-state to state, given action).